// @freshworks/crayons/dist/cjs/fw-toast.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-63c6a574.js');
const toastUtil = require('./toast-util-c28b5264.js');
require('./index-926a762d.js');

const Toast = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    /**
     *  position of the toast notification in screen
     */
    this.position = 'top-center';
    /**
     * Time duration of the toast visibility
     */
    this.timeout = 4000;
    /**
     * Type of the toast - success,failure, warning, inprogress
     */
    this.type = 'warning';
    /**
     *  The Content of the action link
     */
    this.actionLinkText = '';
    /**
     *  won't close automatically
     */
    this.sticky = false;
  }
  componentWillLoad() {
    this.toastContainer = toastUtil.createToastStack(this);
  }
  async trigger(opts) {
    toastUtil.createToastNotification(opts, this.toastContainer, this);
  }
  render() {
    return (index.h(index.Host, null, index.h("slot", null)));
  }
};

exports.fw_toast = Toast;
