// @freshworks/crayons/dist/esm/index.js
export { T as TranslationController } from './Translation-7965186e.js';
import { c as createToastStack, a as createToastNotification } from './toast-util-01ddbcf9.js';
import { c as createProgressLoaderContainer } from './progress-loader-util-1fff4614.js';
import { f as formatDate } from './format-date-util-cbbbafe3.js';
export { r as registerIconLibrary, u as unregisterIconLibrary } from './library.icon.utils-fc805527.js';
import './index-d55b618c.js';
import './index-6f1bd762.js';
import './_commonjsHelpers-9943807e.js';

/** use this file for exposing global functions and ways to set global config **/
function ToastController(config = { position: 'top-center' }) {
  const toastContainer = createToastStack(config);
  function trigger(opts) {
    createToastNotification(opts, toastContainer, config);
  }
  return { trigger };
}
function ProgressLoaderController(config) {
  return createProgressLoaderContainer(config);
}
function DateFormatController({ date, locale, options, } = {
  date: new Date(),
  locale: [],
  options: {},
}) {
  return formatDate({ date, locale, options });
}

export { DateFormatController, ProgressLoaderController, ToastController };
